File: p765/p765.go
package p765

type unionFind struct {
	numOfGroups int
	fa          []int
}

func (uf *unionFind) getfa(cur int) int {
	if uf.fa[cur] == cur {
		return cur
	}

	uf.fa[cur] = uf.getfa(uf.fa[cur])
	return uf.fa[cur]
}

func (uf *unionFind) union(a, b int) {
	x, y := uf.getfa(a), uf.getfa(b)
	if x != y {
		uf.fa[x] = y
		uf.numOfGroups--
	}
}

func minSwapsCouples(row []int) int {
	n := len(row)

	uf := unionFind{}
	uf.numOfGroups = n / 2
	uf.fa = make([]int, n/2)
	for i := 0; i < n/2; i++ {
		uf.fa[i] = i
	}

	for i := 0; i < n; i += 2 {
		coupleIdx1, coupleIdx2 := row[i]/2, row[i+1]/2
		uf.union(coupleIdx1, coupleIdx2)
	}

	return n/2 - uf.numOfGroups
}

File: p765/p765.cpp
// Take all couples as vertex in a graph.

// For even i, if a person from couple u and a person from couple v sit on position i and i+1,
// then connect u and v.

// Suppose finally we have T components and they have n1, n2,...,nT vertex respectively.
// Obviously n1 + n2 + ... + nT = n.
// For each component we only need ni - 1 swaps (Think about permutation!) to make all couples
// in this component holding hands.
// So in total we need n1 - 1 + n2 - 1 + ... + nT - 1 = n - 1 swaps.

class Solution {
public:
	int fa[32];
    int minSwapsCouples(vector<int>& row) {
        int n = row.size()/2;
        int components = n;

        for (int i = 0; i < n; ++i)
        	fa[i] = i;

        for (int i = 0; i < n; ++i)
        {
        	int a = row[2*i]/2, b = row[2*i+1]/2;

        	if (a == b)
        		continue;

        	int u = getFa(a), v = getFa(b);
        	if (u != v)
        	{
        		fa[u] = v;
        		--components;
        	}
        }

        return n - components;
    }

    int getFa(int cur) {
    	if (fa[cur] == cur)
    		return fa[cur];

    	fa[cur] = getFa(fa[cur]);
    	return fa[cur];
    }
};